### ISSUE TITLE
Maybe we need something more obvious for nesting routes

### ISSUE THREAD

**Issue by Aetherall** [faceyspacey/redux-first-router#126]
Hi !
My main ( and only ) problem, is that nesting routes seems not really consistant

let's say I have that kind of app structure

/dashboard  => Dashboard stuff
/dashboard/posts => ListOfPosts
/dashboard/posts/add => AddPostForm while keeping ListOfPosts
/dashboard/posts/:id => EditPostForm while keeping ListOfPosts
  

/dashboard/posts/add =>
````javascript
<Switch>
  Dashboard
  <Switch>
    Posts
    <ListOfPosts />
    <Switch>
      AddPostForm
    </Switch>
  </Switch>
<Switch>
````

The thing is that if I name /dashboard as DASHBOARD and /dashboard/posts as DASHBOARD_POSTS, and then I go to DASHBOARD_POSTS, the matching of the first switch will require additionnal logic to keep it on the right component

I have the feeling that i should treat all my segments as variables and have that kind of routemap
/dashboard(/:1)(/:2)(/:3)

but this would not be consistant as in the future I could add /dashboard/backup or other things which would need another kind of logic...

I may have misunderstood the way to do things though

**Comment by faceyspacey** [comment]
A lot of people have come with this issue. It's primary a result of coming from React Router where there is more automation. So we may either provide components in the future or find some other automated solution.

That said, it's easy to overcome and you will be more in control of your app in terms of explicitness as a result of it. The explicitness will inevitably come in hand as your app grows and you need to customize for new scenarios that don't fit the automagic capabilities standard to React Router style components.

Here's how you do it:

For one you have to think: everything that goes through RFR hits Redux. So that means Redux has knowledge of every route hit, its params as payloads and can conveniently switch over types. Redux is powerful. Anything you can do with Redux, you can do. So the primary thing is related `POST` actions need to trigger a reducer so that its value is something like: `currentList === 'posts'`. And then for the form you might have `currentForm === 'post'`. Now that said, you could just have a `page` reducer or `model` reducer which stores a string corresponding to the name of the current page or model. Then you could have reducers like `showList === true` or `showForm === true`, just as you might have `sidebarOpen === true` or `tabbarVisible === true`. 

You can simplify it all down to one `page` reducer or get more complex, depending on your needs.

Now that said, you can use optional params. They are used the RR4 way with `:foo?` instead of the RR3 way with `(/:foo)`. That will essentially allow you to have less types to switch over in reducers, but at the expense of essentially switching over payload params in your thunk to perform specific thunk actions. I.e. because now you would have less thunks. That decision is up to you. The RFR way there is not to worry about adding more types to switch over in reducers. It's healthy explicitness that will keep you both aware and in control of how different portions of your app affect each other, whereas before Redux apps were prone to inconsistent states because of so many setter/getter style actions. Essentially the contract created by the fact that some actions correspond to some routes results in less actions overall. You have to think that browser/back buttons (or deep-linking in React Native) need to be able to setup the entire app state based on a route. So the contract forces you to make "fatter/smarter" reducers that set themselves up and tare themselves down based on these actions. So you will have less actions overall, but your reducers will have listen to them more to guarantee the correct state. And this is far better than before where you end up doing stupid things like dispatching 3-5 consecutive synchronous actions to create the correct state when a user arrives at important URLs. The easiest example is how easy it would be for a sidebar to end up in an inconsistent state where it was incorrectly open of closed because it wasn't listening to the URL as intelligently as more standard components that make up your route tree. Your components can kind of just sit around and be even dumber, while you move even more business logic to where it belongs: your reducers, aka your domain logic.

Otherwise, make sure to check the top of the readme--a lot of new things have been added, including support for optional params. That may give you some ideas if your hellbent (like some people seem to be) on reducing the # of types switched over in reducers :)

**Comment by faceyspacey** [comment]
Basically the shift is just from listening to the `location` reducer to creating your own reducers (and/or selectors), and listen to them. That's all there is to it.

**Comment by Aetherall** [comment]
Thanks for those explanations ! RFR seems even more interesting to me

In fact I still have some second thoughts about it since react router provide a way to manage the routing far more easily...

Also, why could'nt we manage the routing the same way RR4 does (with components) while keeping it consistent and redux related?
I'm trying to provide to my team a framework where building apps could be as simple as writing react stuff and calling components
I would prefer using RFR for sure, but maybe RR4 is the only way to go?

**Comment by faceyspacey** [comment]
u gotta have sufficient experience with Redux before u can appreciate it :)

U can make the components that operate the same as rr. We just don't have them yet. The plan is to make them mirror the rr API. Just make em. Look at the rr code. They are stupid easy. Then match on the path in Redux location state instead of the context rr uses.

**Comment by faceyspacey** [comment]
basically create the following component and on the line linked replace it with our own match-path package:

https://github.com/ReactTraining/react-router/blob/master/packages/react-router/modules/Route.js#L5

here's our package, which is a fork of their `match-path` functionality:

https://github.com/faceyspacey/rudy-match-path

then instead of getting location state from React Router `context`, you connect the component to redux and get it from `state.location`. Here's what you replace:

https://github.com/ReactTraining/react-router/blob/master/packages/react-router/modules/Route.js#L46

Next, you'll see this line:
https://github.com/ReactTraining/react-router/blob/master/packages/react-router/modules/Route.js#L101

Basically, it's taking the props (i.e. `path`, `exact`, etc) and calling `computePath` which calls `matchPath` with the `nextProps`, and `nextContext.router`. But instead of `nextContext.router`, you have `nextProps.router`, or similar, where you created `router` out of RFR state. So it would be something like: 

```js
this.setState({
      match: this.computeMatch(nextProps, nextProps.router)
})
````

The final component you create can take 1 of 2 routes:

It can mirror the RR Route component api exactly, where each route's component receives , `match`, `location`, `history`, OR you can make it RFR-specific and just pass in `type` and `payload`, or just `state.location`. You won't need `history` anymore since you should be changing the URL now by dispatching actions. You likely won't need all the info in `match` and `location`, as the primary info is just our own `state.location` object. You could also do nothing, and let the child component connect via `react-redux` to extract only the precise state it wants, which would be more optimized to prevent unnecessary re-rendering. 

At the end of the day, you just have to understand the business logic that `<Route />` is doing. It's simply taking 2 paths:

- `<Route path='/foo' />`
- export Route = `connect(state => ({ location: state.location })(RfrRoute)`

Then you need to make:
```js
class RfrRoute extends React.Component {
  // then you do the work here to combine the 2 aspects: redux location state + matching:
  // props.location.pathname + props.path
}
```
> And the `rudy-match-path` library which is pretty much the same as the `matchPath` function imported to RR's Route component will do the work of detecting if the passed path matches the actual path of the address bar in `props.location.pathname`. And if you're not doing `exact`, it will do what you're trying to do where partial paths match, thereby allowing for nested routes. And if you really want to do nested routes fully, make sure you have something like `match.url` as well, which passes in the parent path segment, so you don't have to repeat yourself:

https://reacttraining.com/react-router/core/guides/philosophy/nested-routes

----

So it all boils down to really understanding how this stuff works, which ultimately you will want for any serious project. Most people fly blind. There are a million things to do when building an app, and we often don't have the time, so we have to just *trust* a great number of libraries and their public interface. After all, that's the whole point of a good library. So it's just a matter of whether you really want this, whether it really provides that much value, or whether it's just what you're used to. Basically, if you get to the point where u do all this, you'll basically realize it was unnecessary, and your app would have been more professional had you managed it semi-manually in redux-state + selectors. 

So basically: stop reading this and just use React Router because any bit of time reading this is defeating the purpose and taking value away from React Router. React Router is supposed to be something that requires almost zero thinking. 

However, you'll likely lose lots of time in the long run as it becomes hard to reconcile with Redux where most your state lives. 

Now that said, RFR is not very complicated and only takes a little bit more thinking, but with far more benefits. It's *value* is extremely high, especially for *"real"* apps. If you know how to use Redux, you know how to take advantage of RFR--you know how to free yourself from the limitations of having to reconcile 2 state stores: React Router and its very basic URL state storage + Redux.

**Comment by Aetherall** [comment]
just awesome
Ill go for RFR and create the route component
Many thanks for your answers, they are cristal clear

**Comment by faceyspacey** [comment]
Also keep in mind, there is two ends of the spectrum:

- app specific components that match based on simple reducers like the page and sidebarOpen reducer 
- and advanced React Router style components that match based on non exact paths, thereby enable nesting.

You can make things in between if u want. Ie a simple reusable route component. 

Here's an example for exact matching: u can store in each route on ur routes map a key called "component" which has the actual component class/function to use. And then you simply get it in a parent component like this: routesMap[currentRoute.type].component.

If u want nesting, u could order ur routes so that less specific ones are first and loop through them until one has a path whose indexOf is 0 when compared to the path passed to your custom route component. That's a poor man's nesting. Now that said it isn't necessarily, as u can write any custom logic here to choose the component. A nice way to go is to make a function that takes an action type and perhaps a second arg for kind, ie to indicate u want the correct list. 

So there's a A LOT of options. The takeaway of this comment is that u can reuse the routesMap over and over again and add as many arbitrary key/vals to each route as u want.

**Comment by Aetherall** [comment]
I see, that's great, since I use the addRoute() action creator, 
I could even define in the routes(Map) the name of the corresponding relative file to import in UniversalComponent and keep it consistant in the same time

I'll dive in on monday, and I'll let you know how it's going

**Comment by Aetherall** [comment]
You're a genius that's works perfectly
Thank you you're my hero this week

**Comment by eliseumds** [comment]
Hey @Aetherall, how did you cope with data loading for these nested routes? Do you explicitly, in the thunk related to `/dashboard/posts/abc123`, fetch the list of posts as well? And do you have some logic to short-circuit it when the post ID change?

**Comment by Aetherall** [comment]
@eliseumds 
In my project, I use apollo-client instead of fetching my data manually, so my data is cached and retrieved automatically in an effective way

I suppose you could use redux to store your fetching results so you just have to check whether or not there is already data

**Comment by gotenxds** [comment]
So, I dont know if this was changed in the last six months, but as far as I understand this your suggestion for solving this seemingly simple router task is to attach a function to each route that will parse the route and decide what components should be rendered, this sounds like a router.

Shouldn't the library be the one handling the routing logic ? why is such a simple task that every router should be able to achieve (nesting routes) requires custom logic?

Your are asking users to replace 3 lines of code (rr) with a function per route.

**Comment by faceyspacey** [comment]
That's the react router way. React Router is easy at the easy stuff, but challenging at the real stuff. So the tradeoff is to go a little bit lower level and be really in charge of your *'state'* (i.e. redux state), and have a very manageable situation when it comes to the truly complex task of async data fetching.

At the end of the day async is the only really complex things in modern React apps these days. So we're focused on solving those problems, not the simple problem of conditional logic to pick and choose which component to show.

Also as far as *what a "route" is*, React Router invented and popularized the notion of *routing components*. Before that we used to deal with things in a static map and from here figure out what to render. I.e. in the server-side MVC days, but also in client-routing until React Router came around. ...So perhaps if all you've ever used is React Router, that might be the case. But here we're essentially going back to the "old wisdom" of a "controller layer" that handles the data-fetching, rather than letting it live in the view layer, which can become problematic for a great number of reasons. That's what this router is all about. 

Bottom line: build a big redux + react router + possibly apollo app, and feel the pain of 4 different state stores (you also have local state), then appreciation for having it all under one roof can be gained. Otherwise it's hard to compare the pain of the real problems to the pain of a little work conditionally choosing to what to render from your *own* state.

In the future we'll likely have routing components to help people get started with, as we do realize React Router is so popular that it will help to transfer current understandings (as well as current apps).

Another thing to note: the entire React + Node approach is based on writing more/explicit code to avoid leaky abstractions behind the scenes (yours or 3rd party) breaking our apps without providing much information about what happened. I.e. this is "configuration over convention," whereas the Rails approach was "convention over configuration." So the extra work here in RFR isn't misaligned with general javascript patterns. 

...Now that all said, here's a non-nested route component:

```js
const Route = ({ type }) => components[type]

const components = {
   HOME: Home,
   ABOUT: About
}
```

Obviously you could do it based on params, or types + params. Or any combination of conditional logic. Guess what, that's several hundred lines behind the scenes that React Router is doing to provide you an abstraction you now gotta *trust* where here **you know exactly what you're doing, and no leaks, or misunderstood things, or new package versions conflicts, or perf issues from so much extra code, etc, can get in your way**. The most valuable thing--and again this is part of the JS "configuration" approach--is simply knowing exactly what the code is doing. "Explicitness." 

As we age as programmers, we come to appreciate this more. A lot of the people that find problems with the redux boilerplates are obviously younger programmers exercising there "abstraction" muscles, which is a good thing. But clearly they haven't been burned by the pain of their apps collapsing under the weight of leaky abstractions *yet*. Basically when you're a *learner*, just mastering syntax and all the different language features etc is the goal. So less code is important, as it takes a lot of time to mentally *parse* all that code. That's why they/we all seek to reduce the amount of code at that stage. Code reduction is always important, but it goes down in priority as you progress. Later the goals become different and essentially more complex (with all sorts of layers to it). And what happens is the things you did to keep things DRY actually got in the way of how modular code can be. So more code allows for different layers to the cake you can shift around, whereas if it's all in one grand DRY abstraction, u can't shift it around much, and u begin to make it do so many things that u can't easily reason about that code anymore.

The approach here requires you to make your own little routing-oriented components, but they are simple, non-abstract, and very straightforward. They come in handy when you want to do *that one custom thing* the given library (in this case React Router) can't do--cuz ur not relying on a large abstraction anymore, u can insert the small customization you need right where you need it. Enjoy.

ps. if you're still mastering the react ecosystem, React Router may be the thing to continue with, but if you're getting heavily into redux and really wanna master state-driven principles, RFR is the way to go. No juggling in the view layer, just use react-redux's `connect` to get the state you need.

**Comment by gotenxds** [comment]
"React Router invented and popularized the notion of routing components." This may be correct for react, but this method of routing has been used for a while now and react is not the first one to use it, it has been used by angular, vueJs, emberJs and more.

They all have chosen this method for good reason, it is straightforward, easy to use and understand and has little to no down sides.

As your method does solve the issue of not having two stores in its current state it creates more problems than it solves, to be honest I can't really call it a router as it accepts the user to make the routing by himself.

It works fine for single layer routes, although it requires the user to create multiple maps (type -> path), (type -> componentName), this could be overcome by providing additional meta data to the routesMap object as in ```{path: '', component: ''}```
But this all falls apart when you want to create a more complex app.

RR, angular router, vuejs and many more all handle this by providing the ability to declare "children"
This solves both the path duplication (i.e 'dogs/first', 'dogs/second' -> '/dogs', '/first', '/second') and the issue of identical logic, such as authentication, template rendering and so on.

In the current state of the library (and do correct me if there is a better way) you would need to create yet another map (route -> auth) and use it inside the `onBeforeChange` method.
This is easily handled in nested routes by just specifying the auth logic on the father route.

As an additional note I will state that I'm new to react(a mere few months) but I have been using redux for the last two years and even though I do see the advantages of using RFR, at least for now it seems that the issues far outweighs the benefits.

**Comment by faceyspacey** [comment]
It's all good, don't call it a "router." It's a "controller." 

Basically it boils down to this: you can know your data needs based on the URL + user cookie--it simplifies things a ton to just go straight to the point and get it at a top level all together. **And that's 99.9999% of the time in all apps the world over that you can know and should be able to know your data-needs just by the URL and session token*. Basically the only company in the world that needs component-level data fetching is Facebook where they have so many developers and millions of components, and touching the route is too problematic to just let any old developer do it. 

Therefore that means: for one, we're all basing our decisions on a company we likely will never be like, but secondly the additional indirection of modularizing and doing it at the component level brings lots of extra work and therefore potential pitfalls (as with all *additional indirection*). You better be getting the most out of that modularity if you're going that route--most aren't. 

Here's the problems with the component level data fetching:

- for one, as soon as you have a nested route, that's 2 data-fetches; sequentially
- for 2, if you need to get data based on redux state + routing state + apollo state + local state, you gotta essentially `recompose` 4 hocs. if you're now using the new render prop interfaces to the same features, you're nesting your components until they are unreadable. 
- All of this by the way eventually takes a toll on perf cuz it's a bunch of unnecessary components. I dont care too much about that--just like everyone else who looks the other way on that thought--but i do care about how unreadable and unmaintainable this code is. People's react code is too big.
- now try timing the order of your route change and redux state transition + apollo state
- you're gonna spend your time console.logging lifecycle handlers to see what's triggering them (LOTS OF TIME WASTED). the Redux/RFR approach reduces your components to templates. *component function templates*. period. that's the goal. That's always been the goal. In fact, it's *the dream.* The other approach has the life sucked out of you by lifecycle handlers. Period.
- the list really goes on. The less your components are like templates, the more headaches you have on the horizon. And for what? All in the name of modularized data-fetching? Modularize everything else, just not your async application-specific needs. repeat yourself if you have to, forget DRY. 

Believe me, nested conditional child routes etc etc in terms of rendering is a no brainer. You and I could have made that component in the time we spent talking here. Make it, and own it and be a boss. All the state you need is in Redux.

----

Another thing to note is that the next version has nested routes--but again it's mainly to benefit grouping the fetching of data. That solves the data-acquisition aspects you mentioned. But currently, yes, you create additional maps in global callbacks. 

Now that said, there are also current strategies to reduce duplications, eg: `/dogs/:param?`. that's an optional 2nd parameter there. That's one route. 

So what you're talking about here is this:

```js
const routes { 
    ADMIN: {
         '/dashboard/:param?',
          admin: true
    }
}

onBeforeChange = (dispatch, getState) => {
    const { type } = getState().location
    if (auths[type].admin && !isAllowed(getState()) dispatch(redirect({ type: 'LOGIN' }))
}
```

Your "father" component is gonna have the same logic, except it will be far more error-prone. Getting bent out of shape over adding `admin: true` to route objects is missing the forrest for the trees. The primary problem (the "forrest") is that the `father` component will receive props from other things, and you'll have to deal with that right along side the data-fetching which may or may not be related. You may not even have access to the data you need without additional hocs/renderProps/etc. 

In addition, with the "dynamic routing" approach, u'll have all these children components, and u'll never remember what the rules are for each route. U'll go digging around in your codebase. I really much rather have a map for a small number of routes in a single place, or folder full of them. An app with 20 routes will have 500 components. I don't wanna search through 500 components full of primarily template logic to quickly learn what this app does. Again, you nested your data-fetching needs in components unnecessarily. You almost definitely sacrificed perf to do so. You definitely sacrificed your ability to reason about your code. And at the end of the day, most people end up doing the same sorta thing in components: having a `pages` folder or something similar of your top level routes, similar to how Next.js offers additional data fetching capabilities for their top level components. You're far better off utilizing a route abstraction with callbacks that match its purpose, than callbacks that match the purpose of rendering. It's a better "impedance match" as they say.

----
Another thing: just cuz Angular, Vue, etc are doing it means not a thing brotha man. They're all just copy cats pushing their product as fast as possible, copying the status quo. These guys certainly aren't innovating on the routing front. I would hope not, they have so many other fish to fry. All these view layer routers are the same now. They're clones. And they do almost nothing to help us where we really need help: async, complex route transitions, the zillions of cross-cutting concerns at this intersection like scroll restoration, code-splitting, blocking routes, confirm routes, advanced redirects, the list literally goes on forever. *But they will show a component when the URL matches a path though (reads: underwhelmed)*. They will let you rely on the component host (e.g. React) to put some business logic in lifecycle handlers. NO, as just mentioned, it's not a good *impedance match.*

On a side note, the next version of RFR has its own async middleware where you can specify any callbacks you want, and whether they happen before `enter` or after. You can use sagas, observables or graphql (powered by apollo) instead of thunks. Apollo for example has a middleware plugin that puts the state back into redux properly like you've never seen before. Lastly the router has full control over route cancellation. In fact, it's built on a "commit" model, where ever route change is a request, but not committed until at some point in the middleware it's decided to be committed. All this is to say that there is a huge need for an abstraction other than "components." Something that represents primary state transitions professionally and flexibly. 

Everybody's component obsessed right now brotha, and really there's not much more to components. It's over. New things are comin up. Components are so maxed out on their productivity gains that they need something to deflate them now. Everyone's forced their ware to fit the component paradigm (whether router, data fetching cache, etc), but it's not solving real problems. It's not innovation. It's *adaption*. Sure, it's helpful, however there's major problems in the "component everything" approach I'd like to see solved.

----

Anyway. switch gears. If you've built a large React Native app--which I have--you have the luxury of thinking less in routes, and more in states. Then you essentially can build professionally, as if a router doesnt matter. Thats how this router was born fyi. What happened was: at the end of building that RN app I had to do deep links, emails, etc, so I wanted to make my state-driven app work with URLs. Ie. without changing much code. I was so happy with my code and the architecture (which I had the luxury to obsess over) that I didnt want to sacrifice it for URLs. And so RFR was born. 

> Ask yourself this question: if you didn't have to think about URL paths, would your app be dominated by them? Or would you perhaps entertain some different architecture?? RFR lets you do that.

What we care about is professional state transitions and professional predictable acquisition of data. When you've done that with the modern reactive client-side tools--and im not presuming whether you have or not either way--you gain a really good grasp of how it should be done. What you realize is the URLs (and the way browser back/next buttons work) are dominating the architecture of apps, when they shouldn't be. They're pushing you into a corner, in terms of how you're supposed to do things. All in the name of the URL. 

With this approach, u can nest stuff, and do all sorts of stuff my friend. Just cuz `second` automatically isnt concatenated on to `dog/` *dog* doesn't mean you can't nest things :). Again, we'll provide more automation for that in the next version, but you can architect a pro scheme for all these things in the way RFR has removed the URL's constraints from the equation. **It's just using Redux as if you didn't have to worry about the URL.** That's more power and flexibility of course.

----

As far as the downsides of those routers and the "component everything" approach, there are lots of of them, but the primary one is *predictability*, which also equals testability. **Try time traveling one those apps.** I dont know about you, but I live in the Redux devtools. The Redux/RFR approach has you in control. The "component everything" approach has you fumbling through component lifecycle handlers, wondering what triggered *what*, convoluting your code base in hocs (or render props) to combine states from multiple sources, etc. It's not that I think modular components are bad (obviously they're not), it's about degree, and I KNOW they have gotten out of hand. So RFR brings a little bit of *division of labor* (or whatever the programmer term is for that) in a world that has blown up like a hot air balloon bout to burst. Maybe it was very smart for CSS to become componentetized, but it doesn't mean that it's the case with everything. 

----

Now, in interest of presenting a balanced perspective, I'll share this: the achilles heal of the RFR/Redux approach is modularity--components are more modular than a global state store.  It's why you can't just easily share redux components. To do that, the reducers + actions (sagas, etc) need to be bound to the component. Then you could share it on NPM, and it would just appear in someone's Redux devtools. FYI, this is something I plan to fix in the near future. But it is a problem; it is why components are great. It is the real reason at the core of why Redux is falling out of favor. Apollo is really the spearhead of that, as it has moved away from Redux, and has made components of its own. The core of that of course is the modularity of their data-fetching components. But you can't time travel apollo. Not old or new. It was always half baked and problematic. 

In summary, the real problems are:
- the lack of modularity for Redux
- the additional indirection of components sky-rockets code complexity

If you can put up with making your own route components, and doing some maps in global callbacks, you trade that extra work for an extremely predictable state-transition engine capable of handling async in combination with routing. The developer experience is very good. The architecture is up to you, it's not forced upon you by a back/next browser button (which is the real ring leader here running the show behind the scenes without most us knowing it).
